sslr-core/src/main/java/com/sonar/sslr/impl/typed/SyntaxTreeCreator.java

```diff
@@ -0,0 +1,191 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package com.sonar.sslr.impl.typed;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.GenericTokenType;
+import com.sonar.sslr.api.Token;
+import com.sonar.sslr.api.TokenType;
+import com.sonar.sslr.api.Trivia;
+import com.sonar.sslr.api.Trivia.TriviaKind;
+import com.sonar.sslr.api.typed.ActionParser.GrammarBuilderInterceptor;
+import com.sonar.sslr.api.typed.Input;
+import com.sonar.sslr.api.typed.NodeBuilder;
+import com.sonar.sslr.api.typed.Optional;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+import org.sonar.sslr.internal.grammar.MutableParsingRule;
+import org.sonar.sslr.internal.matchers.ParseNode;
+import org.sonar.sslr.internal.vm.TokenExpression;
+import org.sonar.sslr.internal.vm.TriviaExpression;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.List;
+
+public class SyntaxTreeCreator<T> {
+
+  private final Object treeFactory;
+  private final GrammarBuilderInterceptor mapping;
+  private final NodeBuilder nodeBuilder;
+
+  private final Token.Builder tokenBuilder = Token.builder();
+  private final List<Trivia> trivias = Lists.newArrayList();
+
+  private Input input;
+
+  public SyntaxTreeCreator(Object treeFactory, GrammarBuilderInterceptor mapping, NodeBuilder nodeBuilder) {
+    this.treeFactory = treeFactory;
+    this.mapping = mapping;
+    this.nodeBuilder = nodeBuilder;
+  }
+
+  public T create(ParseNode node, Input input) {
+    this.input = input;
+    this.trivias.clear();
+    T result = (T) visit(node);
+    if (result instanceof AstNode) {
+      ((AstNode) result).hasToBeSkippedFromAst();
+    }
+    return result;
+  }
+
+  private Object visit(ParseNode node) {
+    if (node.getMatcher() instanceof MutableParsingRule) {
+      return visitNonTerminal(node);
+    } else {
+      return visitTerminal(node);
+    }
+  }
+
+  private Object visitNonTerminal(ParseNode node) {
+    MutableParsingRule rule = (MutableParsingRule) node.getMatcher();
+    GrammarRuleKey ruleKey = rule.getRuleKey();
+
+    if (mapping.hasMethodForRuleKey(ruleKey)) {
+      // TODO Drop useless intermediate nodes
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)